applications/blackhole/src/blackhole_socket_callback.erl

```diff
@@ -10,215 +10,94 @@
-        'false' ->
-            lager:info("ahh ahh ahh, you didn't say the magic word")
-    end.
-
--spec build_and_send(kz_json:object()) -> 'ok'.
-build_and_send(Data) ->
-    SendMessage =
-        kz_json:set_values([{<<"Msg-ID">>, kz_util:rand_hex_binary(16)}
-                            | kz_api:default_headers(<<"qubicle">>, <<"1.0">>)
-                           ]
-                          ,Data
-                          ),
-
-    {'ok', Payload} = kz_api:build_message(SendMessage, [], kz_json:get_keys(Data)),
-    amqp_util:basic_publish(<<"qubicle">>, <<"qubicle.recipient">>, Payload, ?DEFAULT_CONTENT_TYPE).
-
-%%--------------------------------------------------------------------
-%% @public
-%% @doc
-%% @end
-%%--------------------------------------------------------------------
--spec close(ne_binary(), ne_binary(), bh_context:context()) -> 'ok'.
-close(SessionPid, SessionId, Context) ->
-    lager:debug("closing socket ~p", [SessionId]),
-    _ = blackhole_tracking:remove_socket(Context),
-    Filter = fun (A, B, C, D) -> filter_bindings(SessionPid, A, B, C, D) end,
+-spec open(any()) -> cb_return().
+open(Ipaddr) ->
+    lager:debug("ws open pid:~p ip:~p", [self(), kz_util:to_binary(inet_parse:ntoa(Ipaddr))]),
+    blackhole_limit:ip(Ipaddr),
+    {'ok', bh_context:new(self())}.
+
+-spec recv(kz_json:object(), bc_context:context()) -> bc_context:context().
+recv(JObj, Context) ->
+    Action = kz_json:get_value(<<"action">>, JObj),
+    Msg = kz_json:delete_key(<<"action">>, JObj),
+    AuthContext = #bh_context{auth_account_id=AuthAccountId} = blackhole_auth:check_auth(Action, Msg, Context),
+    blackhole_limit:account(AuthAccountId),
+    blackhole_limit:message(self()),
+    lager:debug("process_message:~p data:~p auth_account_id:~p", [Action, Msg, AuthAccountId]),
+    process_message(Action, Msg, AuthContext).
+
+-spec close(bh_context:context()) -> 'ok'.
+close(Context = #bh_context{websocket_pid=Pid}) ->
+    blackhole_limit:release(Context),
+    Filter = fun (A, B, C, D) -> filter_bindings(Pid, A, B, C, D) end,
     blackhole_bindings:filter(Filter),
     'ok'.
 
 %%%===================================================================
 %%% Internal functions
 %%%===================================================================
 
-%%--------------------------------------------------------------------
-%% @private
-%% @doc
-%% @end
-%%--------------------------------------------------------------------
--spec maybe_subscribe(bh_context:context(), kz_json:object()) -> cb_return().
--spec maybe_subscribe(bh_context:context(), kz_json:object(), boolean()) -> cb_return().
+-spec process_message(ne_binary(), kz_json:object(), bc_context:context()) -> bc_context:context().
+process_message(<<"subscribe">>, JMsg, Context) -> maybe_subscribe(Context, JMsg);
+process_message(<<"unsubscribe">>, JMsg, Context) -> maybe_unsubscribe(Context, JMsg);
+process_message(<<"authorize">>, _JMsg, Context) ->
```

Comment:
there's a diff between authenticate and authorize. this should be authenticate.  authenticate is a command we handle, authorize is a check we do for the command.